vendor/bat-native-ledger/src/bat/ledger/internal/core/async_result.h

```diff
@@ -0,0 +1,122 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_ASYNC_RESULT_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_CORE_ASYNC_RESULT_H_
+
+#include <list>
+#include <memory>
+#include <utility>
+
+#include "base/callback.h"
+#include "base/optional.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+
+namespace ledger {
+
+// Represents the result of an asynchronous operation.
+//
+// Example:
+//   AsyncResult<int>::Resolver resolver;
+//   resolver.Complete(42);
+//
+//   AsyncResult<int> result = resolver.result();
+//   result.Listen(base::BindOnce([](const int& value) {}));
+//
+// Listeners are called on the current SequencedTaskRunner, and are guaranteed
+// to be called asynchronously. AsyncResult and Resolver objects are internally
+// reference counted and can be passed between sequences; the internal data
+// structures are updated on the sequence that created the Resolver.
+template <typename T>
+class AsyncResult {
+ public:
+  using CompleteType = T;
+  using CompleteCallback = base::OnceCallback<void(const T&)>;
+
+  void Listen(CompleteCallback on_complete) {
+    Listener listener = {.on_complete = std::move(on_complete),
+                         .task_runner = base::SequencedTaskRunnerHandle::Get()};
```

Comment:
This looks like it could very easily be misused to access things on the wrong thread and I'm not seeing the benefit over just passing a callback. cc @iefremov

Reply:
if the purpose of this is to attach multiple callback, I think an observer would be simpler and easier to provide thread guarantees with a thread checker

Reply:
yeah I gather this is something already handled very well using existing tools, like PostTaskAndReplyWithResult. I don't think we shouldn't introduce new async tools without serious reasons

Reply:
Thanks for taking the time to look at this - fortunately we haven't created any PRs yet that use these constructs! 😄 

Ultimately the intent here is to address some of the underlying issues in the `bat-native-ledger` codebase so that rewards devs can be more productive in the future. I'm not particularly attached to any way of doing things but I'm hoping the discussion can proceed with that goal in mind. That said, there is a natural limit to the amount of work that we can invest given other priorities.

**Re threading:** AsyncResult values are always delivered to listeners as const references - an assumption here is that const references will be safe to read concurrently. Obviously you can do things in C++ to make it unsafe.

**Re callbacks:** We have well-known issues in `bat-native-ledger` with the current callback approach. It is natural for devs to want to compose async tasks out of smaller async tasks, but this has lead to readability problems for the codebase, where callbacks are bound in to other callbacks (which are bound in to other callbacks, etc.). AsyncResult is one part of an intended solution out of this mess (the other being "Task" components).

**Re observers:** Defining observers for each abstraction that does something asynchronous is another option for dealing with the callback problem that we currently have, and one that I considered. I think it is appropriate when a component does things "on its own" (e.g. on a timer) and needs to publish those events. I think it is less appropriate for managing async "flow" tasks (of which there are many in `bat-native-ledger`); using the observer pattern in those cases also leads to code understandability problems.

**Re existing tools:** Trust me, I don't introduce new primitives lightly (and would prefer not to at all), but as mentioned above we have some issues with the current codebase and the existing tools haven't helped. Also, one of the goals is to take something that we must surely admit is fairly painful in Chromium code (async flows), and make it simpler for devs.

Reply:
I guess existing tools haven't helped because nobody tried to use them? I'd say Chromium async tools are pretty much straightforward when you get used to them.

just for clarity: how often would we need multiple listeners for a given task? Ho often do they listen on different sequences?

Reply:
Happy to learn more!

> how often would we need multiple listeners for a given task

It comes up, in particular if you want to dedupe tasks. When I did the publisher info fetcher, you can see [here](https://github.com/brave/brave-core/blob/master/vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.h#L61) how I had to put together a map of `publisher_key -> vector<callback>` and manage all that. Of course, it would have been a bit easier with `base::CallbackList`, but we couldn't do that because `bat-native-ledger` doesn't use `Bind`. (Part of what I'm trying to change!) It would have been even easier with AsyncResult, though, because I could have just stored a map of `publisher_key -> AsyncResult<T>` and handled those out to consumers as they requested publisher info.

> Ho often do they listen on different sequences?

Not sure - I added the sequence logic in later after I learned a bit more about how sequences work and how I would want AsyncResult to work if there's more than one sequence. (I could very well not have something correct, though.) It's not a requirement for `bat-native-ledger` right now because that library only uses a single sequence.

In general, "reifying" the result of an async operation as a first class value that you can pass around means that you can abstract over it in a generic way. So for instance, you can write something like `context()->StartTask<MyTask>(...)` or a generic mapping function and things like that.

Reply:
maybe we could start with transitioning from std::bind to base::Bind first

it would be nice to finally understand how many sequenced are used by Ledger. E.g., I see `task_runner_` in `LedgerImpl` that is apparently not used 🤔